letsencrypt/client/client.py

```diff
@@ -389,18 +345,57 @@ def init_csr(privkey, names, cert_dir):
 
     return le_util.CSR(csr_filename, csr_der, "der")
 
+
 # This should be controlled by commandline parameters
-def determine_authenticator(config):
+def determine_authenticator(all_auths):
     """Returns a valid IAuthenticator.
 
-    :param config: Configuration.
-    :type config: :class:`letsencrypt.client.interfaces.IConfig`
+    :param list all_auths: Where each is a tuple of the form
+        ('description', 'IAuthenticator', *options..) where IAuthenticator is a
+        :class:`letsencrypt.client.interfaces.IAuthenticator` object or class
+        and options are the parameters used to initialize the authenticator.
+
+    :returns: Valid Authenticator object or None
 
     """
-    try:
-        return configurator.ApacheConfigurator(config)
-    except errors.LetsEncryptNoInstallationError:
-        logging.info("Unable to determine a way to authenticate the server")
+    # Available Authenticator objects
+    avail_auths = []
+    # Error messages for misconfigured authenticators
+    errs = {}
+
+    for pot_auth in all_auths:
+        try:
+            # I do not think this a great solution but haven't come up with
+            # anything better yet... other than constricting init functions for
+            # authenticators
```

Comment:
Mostly because I wanted to differentiate "misconfigured" (but technically available) authenticators to not available authenticators.  The logic to instantiate them needs to go somewhere and I would rather pass all of the known authenticators to the function and let the function decide how to proceed.  (We should display the class as available (it is installed) but not allow them to choose it... while explaining the reason).

The alternative is to have an extremely minimal init function for all authenticator/installers and then have a separate sanity check function.  (I was trying very hard to avoid this as it complicates the actual plugins).  This would also solve the description problem that I will describe at the appropriate comment.

I did mention possibly constricting the init function to only handle particular parameters.  This could lead to an incredibly bloated config object, but this may be the route to go.

The plugin system based on setuptools entry points sounds good.